core/liquidity/v2/snapshot.go

```diff
@@ -0,0 +1,461 @@
+}
+
+func (e *SnapshotEngine) serialise(k string) ([]byte, error) {
+	var (
+		buf     []byte
+		changed bool
+		err     error
+	)
+
+	switch k {
+	case e.provisionsKey:
+		buf, changed, err = e.serialiseProvisions()
+	case e.performancesKey:
+		buf, changed, err = e.serialisePerformances()
+	case e.suppliedKey:
+		buf, changed, err = e.serialiseSupplied()
+	case e.scoresKey:
+		buf, changed, err = e.serialiseScores()
+	default:
+		return nil, types.ErrSnapshotKeyDoesNotExist
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if e.stopped {
+		return nil, nil
+	}
+
+	if !changed {
+		return buf, nil
+	}
+
+	switch k {
+	case e.provisionsKey:
+		e.serialisedProvisions = buf
+	case e.performancesKey:
+		e.serialisedPerformances = buf
+	case e.suppliedKey:
+		e.serialisedSupplied = buf
+	case e.scoresKey:
+		e.serialisedScores = buf
+	default:
+		return nil, types.ErrSnapshotKeyDoesNotExist
+	}
+
+	return buf, nil
+}
+
+func (e *SnapshotEngine) serialiseProvisions() ([]byte, bool, error) {
+	// these are sorted already, only a conversion to proto is needed
+	lps := e.Engine.provisions.Slice()
+	pblps := make([]*typespb.LiquidityProvision, 0, len(lps))
+	for _, v := range lps {
+		pblps = append(pblps, v.IntoProto())
+	}
+
+	payload := &snapshotpb.Payload{
+		Data: &snapshotpb.Payload_LiquidityV2Provisions{
+			LiquidityV2Provisions: &snapshotpb.LiquidityV2Provisions{
+				MarketId:            e.market,
+				LiquidityProvisions: pblps,
+			},
+		},
+	}
+
+	return e.marshalPayload(payload)
+}
+
+func (e *SnapshotEngine) serialisePerformances() ([]byte, bool, error) {
+	// Extract and sort the parties to serialize a deterministic array.
+	parties := make([]string, 0, len(e.slaPerformance))
+	for party := range e.slaPerformance {
+		parties = append(parties, party)
+	}
+	sort.Strings(parties)
+
+	performancePerPartySnapshot := make([]*snapshotpb.LiquidityV2PerformancePerParty, 0, len(e.slaPerformance))
+	for _, party := range parties {
+		partyPerformance := e.slaPerformance[party]
+
+		registeredPenaltiesPerEpochSnapshot := make([]string, 0, len(partyPerformance.allEpochsPenalties))
+		for _, registeredPenalty := range partyPerformance.allEpochsPenalties {
+			registeredPenaltiesPerEpochSnapshot = append(registeredPenaltiesPerEpochSnapshot, registeredPenalty.String())
+		}
+
+		partyPerformanceSnapshot := &snapshotpb.LiquidityV2PerformancePerParty{
+			Party:                            party,
+			ElapsedTimeMeetingSlaDuringEpoch: partyPerformance.s.String(),
+			CommitmentStartTime:              partyPerformance.start.String(),
+			RegisteredPenaltiesPerEpoch:      registeredPenaltiesPerEpochSnapshot,
+		}
+		performancePerPartySnapshot = append(performancePerPartySnapshot, partyPerformanceSnapshot)
+	}
+
+	payload := &snapshotpb.Payload{
+		Data: &snapshotpb.Payload_LiquidityV2Performances{
+			LiquidityV2Performances: &snapshotpb.LiquidityV2Performances{
+				MarketId:            e.market,
+				EpochStartTime:      e.slaEpochStart.String(),
```

Comment:
For saving times in snapshots we usually just save the `.UnixNano()`. We should do that here too 
(the docs for `time.String()` say "The returned string is meant for debugging" and mention stuff about not using it for stable serialization)

Reply:
Oh, right.